### PR TITLE
cache: add fetch function

### PR DESCRIPTION
**Pull request by sergeylanzman**

### CODE REVIEW

File: cache/cache.go [hunk 1]
```diff
@@ -49,6 +49,8 @@ import (
 type Cache interface {
 	// get cached value by key.
 	Get(key string) interface{}
+	// fetch cached value by key if not exist generate new cache payload.
```
Comment (LasTshaMAN) [inline cache/cache.go]: if not exist -> if does not exist

Comment (LasTshaMAN) [inline cache/cache.go]: Not very clear from this comment what this function actually accomplishes.

Comment (astaxie) [inline cache/cache.go]: yea, I also don't think this is necessary. it looks like the hook for the Get.

Comment (sergeylanzman) [inline cache/cache.go]: yea this not very clear. You have other ideas?

Comment (eyalpost) [inline cache/cache.go]: getOrCreate?

Comment (astaxie) [inline cache/cache.go]: Yea, GetOrCreate sounds good to me

File: cache/cache_test.go [hunk 2]
```diff
@@ -95,6 +95,26 @@ func TestCache(t *testing.T) {
...
+		t.Error("fetch error")
+	}
+	if bm.Get(key).(string) != "payload" {
+		t.Error("get err")
+	}
+	time.Sleep(11 * time.Second)
+	if bm.Get(key) != nil {
+		t.Error("get err")
+	}
+	payload = bm.Fetch(key, timeoutDuration, genfunc)
```
Comment (gogitops) [inline cache/cache_test.go:114]: warning: ineffectual assignment to payload (ineffassign)

File: cache/cache_test.go [hunk 3]
```diff
@@ -164,5 +184,25 @@ func TestFileCache(t *testing.T) {
...
+		t.Error("fetch error")
+	}
+	if bm.Get(key).(string) != "payload" {
+		t.Error("get err")
+	}
+	time.Sleep(11 * time.Second)
+	if bm.Get(key).(string) != "" {
+		t.Error("get err")
+	}
+	payload = bm.Fetch(key, timeoutDuration, genfunc)
```
Comment (gogitops) [inline cache/cache_test.go:202]: warning: ineffectual assignment to payload (ineffassign)

File: cache/memcache/memcache_test.go [hunk 4]
```diff
@@ -101,6 +101,26 @@ func TestMemcacheCache(t *testing.T) {
...
+		t.Error("fetch error")
+	}
+	if string(bm.Get(key).([]byte)) != "payload" {
+		t.Error("get err")
+	}
+	time.Sleep(11 * time.Second)
+	if bm.Get(key) != nil {
+		t.Error("get err")
+	}
+	payload = bm.Fetch(key, timeoutDuration, genfunc)
```
Comment (gogitops) [inline cache/memcache/memcache_test.go:119]: warning: ineffectual assignment to payload (ineffassign)

File: cache/ssdb/ssdb_test.go [hunk 5]
```diff
@@ -94,6 +94,26 @@ func TestSsdbcacheCache(t *testing.T) {
...
+		t.Error("fetch error")
+	}
+	if ssdb.Get(key).(string) != "payload" {
+		t.Error("get err")
+	}
+	time.Sleep(11 * time.Second)
+	if ssdb.Get(key) != nil {
+		t.Error("get err")
+	}
+	payload = ssdb.Fetch(key, timeoutDuration, genfunc)
```
Comment (gogitops) [inline cache/ssdb/ssdb_test.go:112]: warning: ineffectual assignment to payload (ineffassign)

**Pull request closed without merging**